exporter/splunkhecexporter/logdata_to_splunk.go

```diff
@@ -15,29 +15,156 @@
+
+	go func() {
+		defer close(eventsCh)
+
+		// event buffers a single event.
+		event := new(bytes.Buffer)
+		encoder := json.NewEncoder(event)
+
+		// events buffers events up to the max content length.
+		events := &eventsBuf{buf: new(bytes.Buffer)}
+
+		rl := ld.ResourceLogs()
+		for i := 0; i < rl.Len(); i++ {
+			ill := rl.At(i).InstrumentationLibraryLogs()
+			for j := 0; j < ill.Len(); j++ {
+				l := ill.At(j).Logs()
+				for k := 0; k < l.Len(); k++ {
+					select {
+					case <-ctx.Done():
+						return
+					default:
+						if err := encoder.Encode(mapLogRecordToSplunkEvent(l.At(k), config, logger)); err != nil {
+							eventsCh <- &eventsBuf{buf: nil, index: nil, err: consumererror.Permanent(err)}
+							return
+						}
+						event.WriteString("\r\n\r\n")
+
+						// The size of an event must be less than or equal to max content length.
+						if config.MaxContentLength > 0 && event.Len() > config.MaxContentLength {
+							err := fmt.Errorf("found a log event bigger than max content length (event: %d bytes, max: %d bytes)", config.MaxContentLength, event.Len())
+							eventsCh <- &eventsBuf{buf: nil, index: nil, err: consumererror.Permanent(err)}
+							return
+						}
+
+						// Moving the event to events.buf if length will be <= max content length.
+						// Max content length <= 0 is interpreted as unbound.
+						if events.buf.Len()+event.Len() <= config.MaxContentLength || config.MaxContentLength <= 0 {
+							// WriteTo() empties and resets buffer event.
+							if _, err := event.WriteTo(events.buf); err != nil {
+								eventsCh <- &eventsBuf{buf: nil, index: nil, err: consumererror.Permanent(err)}
+								return
+							}
+
+							// Setting events index using the log record indices of the 1st event.
+							if events.index == nil {
+								events.index = &eventIndex{src: i, lib: j, log: k}
+							}
+
+							continue
+						}
+
+						eventsCh <- events
+
+						// Creating a new events buffer.
+						events = &eventsBuf{buf: new(bytes.Buffer)}
+						// Setting events index using the log record indices of any current leftover event.
+						if event.Len() != 0 {
+							events.index = &eventIndex{src: i, lib: j, log: k}
+						}
+					}
+				}
+			}
+		}
+
+		// Writing any leftover event to eventsBuf buffer `events.buf`.
+		if _, err := event.WriteTo(events.buf); err != nil {
+			eventsCh <- &eventsBuf{buf: nil, index: nil, err: consumererror.Permanent(err)}
+			return
+		}
+
+		eventsCh <- events
+
+	}()
+	return eventsCh, cancel
+}
+
+func (ld *logDataWrapper) countLogs(start *eventIndex) int {
+	count, orig := 0, *ld.InternalRep().Orig
+	for i := start.src; i < len(orig); i++ {
+		for j, iLLogs := range orig[i].InstrumentationLibraryLogs {
+			switch {
+			case i == start.src && j < start.lib:
+				continue
+			default:
+				count += len(iLLogs.Logs)
 			}
 		}
 	}
+	return count - start.log
+}
+
+func (ld *logDataWrapper) trimLeft(end *eventIndex) *pdata.Logs {
+	clone := ld.Clone()
```

Comment:
Are we going to have a bit more memory usage because we're cloning and manipulating data structures here?

Could we just keep currentIndex a field of logDataWrapper and keep incrementing it until all records are parsed?